contributors/design-proposals/bulk_watch.md

```diff
@@ -0,0 +1,348 @@
+secret may also lose access to it and this watch should be closed in this case.
+
+ We would like the API to be generic enough so that it will be useful for many
+different usecases, not just for this particular one. As a result, below we are
+describing requirements that the solution has to meet to satisfy our needs:
+- a single bulk request has to support multiple resources (e.g. get a node and
+all pods associated with it)
+- if the bulk request filters asks only for object of a single resource type,
+the returned result should ideally be comparible with what is returned from
+the regular API
+- the API has to support aggregation if different resource types are server by
+different apiservers
+- ACLs of watches have to be periodically verified and automatically closed if
+ACLs no longer allow for this watch.
+- at some point in the future we want clients to be able to alter their watch
+subscriptions incrementally
+
+
+# Detailed design
+
+ To make the API generic enough, one of our requirements is to make those bulk
+operations work across different resource types (e.g. watch pod P and secret S
+within a single watch call). [Though, this doesn't need to be supported by the
+initial implementation - this just have to be possible at the API level.]
+
+ Spanning multiple resources, resource types or conditions will be increasingly
+important for large number of watches. As an example, federation will add more
+watches for every type it federates. With that in mind, bypassing resource type
+level aggregation and going to aggregation over resource types will allow us to
+more aggresively optimize in the future. More
+
+ Moreover, our current REST API doesn't offer an easy way to handle "multiple
+watches of a given type". As a result, instead of inventing new top level
+pattern per type, we can introduce a new resource type that follows normal
+RESTful rules and solves even more generic problem os spanning multiple
+different resource types.
+
+ We will introduce this new resource in a new dedicated API group:
+```
+/apis/bulk.k8s.io/
+```
+ that underneath will have a separate implementation. Initially we are going
+to support the following operations:
+```
+POST /apis/bulk.k8s.io/v1/get   <body defines filtering>
+```
+ Similarly as in our main api, we will use the same resource for serving both
+LIST and WATCH, and those will differ by appropriately setting `?watch=1 `
+parameter in the request.
+
+ Later, we would like to add ability for bulk create/update/delete with a
+similar pattern, though it's not part of this design.
+
+
+### Filtering definition
+
+ We need to make our filtering mechanism to support different resource types at
+the same time. On the other hand, we would like it to be as consistent with all
+other Kubernetes APIs as possible. So we define the selector for bulk operations
+as following:
+
+```
+// TODO: Find a better name for it?
+type BulkSelector struct {
```

Comment:
Rather than normalizing this to two levels, why not have only one level which is

```
type Bulk struct {
  Target []ListSelector
}
type ListSelector struct {
  Resource GroupVersionResource
  ListOptions
}